File: scripts/gui/networks.lua
local priorities = require "scripts.gui.priorities"
local sig_lib = require("__cybersyn2-combinator__/lib/core/signal")

local networks = {}

function networks.are_networks_equal(old_list, new_list)
  if old_list == new_list then return true end
  if not old_list or not new_list then return false end
  if #old_list ~= #new_list then return false end
  for i = 1, #new_list do
    local a = old_list[i]
    local b = new_list[i]
    if not a or not b or not a.Signal or not b.Signal then return false end
    if a.Signal.name ~= b.Signal.name or a.Signal.type ~= b.Signal.type or a.Signal.quality ~= b.Signal.quality or a.Count ~= b.Count then
      return false
    end
  end
  return true
end

function networks.get_all_global_active_networks()
  if not game or not remote.interfaces["cybersyn2"] then return {} end

  local unique_map = {}

  local process_networks_dict = function(dict)
    if not dict or type(dict) ~= "table" then return end
    for k, count in pairs(dict) do
      local parsed = priorities.parse_signal_key(k)
      if not parsed then
        local stype = sig_lib.get_signal_type_from_name(k)
        if stype then parsed = { type = stype, name = k } end
      end
      if parsed and parsed.name then
        local c = tonumber(count) or 0
        local key = (parsed.type or "item") .. "_" .. parsed.name .. "_" .. (parsed.quality or "normal") .. "_" .. tostring(c)
        if not unique_map[key] then
          unique_map[key] = { Signal = parsed, Count = c }
        end
      end
    end
  end

  local process_networks_on_object = function(obj)
    local net = obj.networks
    if not net or type(net) ~= "table" or not next(net) then
      net = obj.network or obj.network_mask or obj.network_flag or obj.network_id
    end
    process_networks_dict(net)
  end

  local stops_res = remote.call("cybersyn2", "query", { type = "stops", all = true })
  if not stops_res or not stops_res.data then return {} end

  local inv_ids = {}
  for _, stop in ipairs(stops_res.data) do
    process_networks_on_object(stop)
    if stop.inventory_id then table.insert(inv_ids, stop.inventory_id) end
    if stop.shared_inventory_master then table.insert(inv_ids, stop.shared_inventory_master) end
  end

  if #inv_ids > 0 then
    local inv_res = remote.call("cybersyn2", "query", { type = "inventories", ids = inv_ids })
    if inv_res and inv_res.data then
      for _, inv in ipairs(inv_res.data) do
        if inv.orders then
          for _, order_obj in ipairs(inv.orders) do
            process_networks_on_object(order_obj)
          end
        end
      end
    end
  end

  local result = {}
  for _, entry in pairs(unique_map) do
    table.insert(result, entry)
  end
  table.sort(result, function(a, b)
    local na = a.Signal.name or ""
    local nb = b.Signal.name or ""
    if na ~= nb then return na < nb end
    return (a.Count or 0) < (b.Count or 0)
  end)
  return result
end

return networks

File: src/scripts/gui/networks.ts
import { parseSignalKey } from './priorities';
import * as sigLib from 'fcore/utils/signal';

export function areNetworksEqual(
  oldList: ActiveNetworkEntry[] | undefined,
  newList: ActiveNetworkEntry[] | undefined,
): boolean {
  if (oldList === newList) return true;
  if (!oldList || !newList) return false;
  if (oldList.length !== newList.length) return false;
  for (let i = 0; i < newList.length; i++) {
    const a = oldList[i];
    const b = newList[i];
    if (!a || !b || !a.signal || !b.signal) return false;
    if (
      a.signal.name !== b.signal.name ||
      a.signal.type !== b.signal.type ||
      a.signal.quality !== b.signal.quality ||
      a.count !== b.count
    ) {
      return false;
    }
  }
  return true;
}

export function getAllGlobalActiveNetworks(): ActiveNetworkEntry[] {
  if (!game || !remote.interfaces['cybersyn2']) return [];

  const uniqueMap: Record<string, ActiveNetworkEntry> = {};

  const processNetworksDict = (dict: any) => {
    if (!dict || typeof dict !== 'object') return;
    for (const [k, count] of pairs(dict as Record<string, number>)) {
      let parsed = parseSignalKey(k as string);
      if (!parsed) {
        const stype = sigLib.getSignalTypeFromName(k as string);
        if (stype) parsed = { type: stype, name: k as string };
      }
      if (parsed && parsed.name) {
        const c = tonumber(count) || 0;
        const key = `${parsed.type || 'item'}_${parsed.name}_${parsed.quality || 'normal'}_${c}`;
        if (!uniqueMap[key]) {
          uniqueMap[key] = { signal: parsed, count: c };
        }
      }
    }
  };

  const processNetworksOnObject = (obj: any) => {
    let net = obj.networks;
    if (!net || typeof net !== 'object' || next(net)[0] === undefined) {
      net = obj.network || obj.network_mask || obj.network_flag || obj.network_id;
    }
    processNetworksDict(net);
  };

  const stopsRes = remote.call('cybersyn2', 'query', { type: 'stops', all: true }) as any;
  if (!stopsRes || !stopsRes.data) return [];

  const invIds: number[] = [];
  for (const stop of stopsRes.data) {
    processNetworksOnObject(stop);
    if (stop.inventory_id) invIds.push(stop.inventory_id);
    if (stop.shared_inventory_master) invIds.push(stop.shared_inventory_master);
  }

  if (invIds.length > 0) {
    const invRes = remote.call('cybersyn2', 'query', { type: 'inventories', ids: invIds }) as any;
    if (invRes && invRes.data) {
      for (const inv of invRes.data) {
        if (inv.orders) {
          for (const orderObj of inv.orders) {
            processNetworksOnObject(orderObj);
          }
        }
      }
    }
  }

  const result: ActiveNetworkEntry[] = [];
  for (const [_, entry] of pairs(uniqueMap)) {
    result.push(entry);
  }
  result.sort((a, b) => {
    const na = a.signal.name || '';
    const nb = b.signal.name || '';
    if (na !== nb) return na < nb ? -1 : 1;
    return (a.count || 0) - (b.count || 0);
  });
  return result;
}
